exercises/csharp/add-esdk-complete/src/Api.cs

```diff
@@ -0,0 +1,161 @@
+            PointerItem pointerItem = PointerItem.FromItem(response.Item);
+            return pointerItem;
+        }
+
+        protected async Task<PointerItem> GetPointerItem(ContextItem contextItem)
+        {
+            return await GetPointerItem(contextItem.SortKey.S);
+        }
+
+        protected async Task<HashSet<PointerItem>> QueryForContextKey(string contextKey)
+        {
+            QueryRequest request = ContextItem.QueryFor(contextKey);
+            request.TableName = tableName;
+            QueryResponse response = await amazonDynamoDBClient.QueryAsync(request);
+            HashSet<ContextItem> contextItems = new(response.Items.Select(i => ContextItem.FromItem(i)));
+            var pointerItems = await Task.WhenAll(contextItems.Select(async i => await GetPointerItem(i)));
+            return new HashSet<PointerItem>(pointerItems);
+        }
+
+        protected async Task WriteObject(DocumentBundle bundle)
+        {
+            byte[] data = bundle.Data;
+            //metadata
+            var request = new PutObjectRequest
+            {
+                BucketName = bucketName,
+                Key = bundle.Pointer.PartitionKey.S,
+            };
+            foreach (KeyValuePair<string, string> kvp in bundle.Pointer.GetContext())
+            {
+                request.Metadata.Add(kvp.Key, kvp.Value);
+            }
+            request.Headers.ContentLength = data.Length;
+            request.InputStream = new MemoryStream(data);
+            await amazonS3Client.PutObjectAsync(request);
+        }
+
+        protected async Task<byte[]> GetObjectData(string key)
+        {
+            using (GetObjectResponse response = await amazonS3Client.GetObjectAsync(bucketName, key))
+            {
+
+                using (var memoryStream = new MemoryStream())
+                {
+                    response.ResponseStream.CopyTo(memoryStream);
+                    return memoryStream.ToArray();
+                }
+            }
+        }
+
+        public async Task<HashSet<PointerItem>> List()
+        {
+            ScanResponse response = await amazonDynamoDBClient.ScanAsync(tableName, PointerItem.FilterFor());
+            return new HashSet<PointerItem>(response.Items.Select(i => PointerItem.FromItem(i)));
+        }
+
+        public async Task<PointerItem> Store(byte[] data)
+        {
+            return await Store(data, new());
+        }
+
+        public async Task<PointerItem> Store(byte[] data, Dictionary<string, string> context)
+        {
+            // ADD-ESDK-START: Add Encryption to store
+            var encryptedMessage = awsEncryptionSdk.Encrypt(new EncryptInput
+            {
+                Plaintext = new MemoryStream(data),
+                Keyring = keyring
+            });
+
+            DocumentBundle bundle = DocumentBundle.FromDataAndContext(encryptedMessage.Ciphertext.ToArray(), context);
+            await WriteItem(bundle.Pointer);
+            await WriteObject(bundle);
```

Comment:
nit: is it necessary to `await` these in order, or can they be awaited concurrently?